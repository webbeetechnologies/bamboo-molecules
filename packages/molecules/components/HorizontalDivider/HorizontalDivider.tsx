import { memo, useMemo } from 'react';
import { type ViewStyle, type StyleProp, type ViewProps, View } from 'react-native';
import { StyleSheet } from 'react-native-unistyles';

import { registerComponentStyles, getRegisteredMoleculesComponentStyles } from '../../core';

export type Props = Omit<ViewProps, 'children'> & {
    /**
     * left inset of the divider.
     */
    leftInset?: number;
    /**
     * right inset of the divider.
     */
    rightInset?: number;
    /**
     *  Whether divider should be bolded.
     */
    bold?: boolean;
    /**
     *  Vertical spacing of the Divider
     */
    spacing?: number;
    style?: StyleProp<ViewStyle>;
};

/**
 * A divider is a thin, lightweight separator that groups content in lists and page layouts.
 *
 * <div class="screenshots">
 *  <figure>
 *    <img class="medium" src="screenshots/divider.png" />
 *  </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { View } from 'react-native';
 * import { Divider, Text } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <View>
 *     <Text>Lemon</Text>
 *     <Divider />
 *     <Text>Mango</Text>
 *     <Divider />
 *   </View>
 * );
 *
 * export default MyComponent;
 * ```
 */

const HorizontalDivider = ({
    leftInset = 0,
    rightInset = 0,
    style,
    bold = false,
    spacing = 0,
    ...rest
}: Props) => {
    horizontalDividerStyles.useVariants({
        isBold: bold,
    });

    const memoizedStyles = useMemo(() => {
        return [
            horizontalDividerStyles.root,
            leftInset && { marginLeft: leftInset },
            rightInset && { marginRight: rightInset },
            spacing && { marginVertical: spacing },
            style,
        ];
    }, [leftInset, rightInset, spacing, style]);

    return <View {...rest} style={memoizedStyles as StyleProp<ViewStyle>} />;
};

export const horizontalDividerStylesDefault = StyleSheet.create(theme => ({
    root: {
        height: StyleSheet.hairlineWidth,
        background: theme.colors.outlineVariant,

        variants: {
            isBold: {
                true: {
                    height: 1,
                },
            },
        },
    },
}));

registerComponentStyles('HorizontalDivider', horizontalDividerStylesDefault);
export const horizontalDividerStyles = getRegisteredMoleculesComponentStyles('HorizontalDivider');

export default memo(HorizontalDivider);
