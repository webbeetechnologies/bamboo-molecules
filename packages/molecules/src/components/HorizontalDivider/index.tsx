export {
    default as HorizontalDivider,
    Props as HorizontalDividerProps,
    horizontalDividerStyles,
} from './HorizontalDivider';
