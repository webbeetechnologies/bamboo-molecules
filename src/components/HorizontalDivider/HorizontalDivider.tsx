import { memo } from 'react';
import { StyleSheet, ViewStyle, StyleProp } from 'react-native';
import type { ViewProps } from '@webbee/bamboo-atoms';
import { useComponentStyles, useMolecules } from '../../hooks';
import { withNormalizedStyleProp } from '../../hocs';

export type Props = Omit<ViewProps, 'children'> & {
    /**
     * left inset of the divider.
     */
    leftInset?: number;
    /**
     * right inset of the divider.
     */
    rightInset?: number;
    /**
     *  Whether divider should be bolded.
     */
    bold?: boolean;
    /**
     *  Vertical spacing of the Divider
     */
    spacing?: number;
    style?: StyleProp<ViewStyle>;
};

/**
 * A divider is a thin, lightweight separator that groups content in lists and page layouts.
 *
 * <div class="screenshots">
 *  <figure>
 *    <img class="medium" src="screenshots/divider.png" />
 *  </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { View } from 'react-native';
 * import { Divider, Text } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <View>
 *     <Text>Lemon</Text>
 *     <Divider />
 *     <Text>Mango</Text>
 *     <Divider />
 *   </View>
 * );
 *
 * export default MyComponent;
 * ```
 */

const HorizontalDivider = ({
    leftInset = 0,
    rightInset = 0,
    style,
    bold = false,
    spacing = 0,
    ...rest
}: Props) => {
    const { View } = useMolecules();
    const { bold: boldStyles, ...defaultStyles } = useComponentStyles('HorizontalDivider', style);

    return (
        <View
            {...rest}
            style={[
                defaultStyles,
                leftInset && { marginLeft: leftInset },
                rightInset && { marginRight: rightInset },
                bold && boldStyles,
                spacing && { marginVertical: spacing },
            ]}
        />
    );
};

export const horizontalDividerStyles = {
    height: StyleSheet.hairlineWidth,
    background: 'colors.surfaceVariant',
    bold: {
        height: 1,
    },
};

export default memo(withNormalizedStyleProp(HorizontalDivider));
