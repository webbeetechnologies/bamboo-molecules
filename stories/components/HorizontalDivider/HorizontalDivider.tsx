import { useMolecules } from '../../../src';
import { HorizontalDivider, HorizontalDividerProps } from '../../../src/components';

export type Props = HorizontalDividerProps & {};

export const Example = (props: Props) => {
    const { View, Text } = useMolecules();

    return (
        <View>
            <Text>Above</Text>
            <HorizontalDivider spacing={10} {...props} />
            <Text>Below</Text>
        </View>
    );
};
