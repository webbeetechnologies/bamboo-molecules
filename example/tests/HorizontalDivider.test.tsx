import { Example as HorizontalDivider } from '../stories/components/HorizontalDivider/HorizontalDivider';
import { renderWithWrapper } from '../testHelper';

it('should render VerticalDivider', () => {
    const tree = renderWithWrapper(<HorizontalDivider />).toJSON();
    expect(tree).toMatchSnapshot();
});

it('should render VerticalDivider with spacing of 10 pixels', () => {
    const tree = renderWithWrapper(<HorizontalDivider spacing={10} />).toJSON();
    expect(tree).toMatchSnapshot();
});
