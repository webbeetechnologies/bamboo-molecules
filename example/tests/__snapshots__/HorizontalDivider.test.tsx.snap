// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render VerticalDivider 1`] = `
<View
  style={Object {}}
>
  <Text
    style={Object {}}
  >
    Above
  </Text>
  <View
    style={
      Object {
        "background": "rgba(231, 224, 236, 1)",
        "height": 0.5,
        "marginVertical": 10,
      }
    }
  />
  <Text
    style={Object {}}
  >
    Below
  </Text>
</View>
`;

exports[`should render VerticalDivider with spacing of 10 pixels 1`] = `
<View
  style={Object {}}
>
  <Text
    style={Object {}}
  >
    Above
  </Text>
  <View
    style={
      Object {
        "background": "rgba(231, 224, 236, 1)",
        "height": 0.5,
        "marginVertical": 10,
      }
    }
  />
  <Text
    style={Object {}}
  >
    Below
  </Text>
</View>
`;
